### ISSUE TITLE
TypeError : strlen(): Argument #1 ($string) must be of type string, bool given in FHIRTiming

### ISSUE THREAD

**Issue by leonrenkema** [dcarbone/php-fhir#82]
I am getting the following error `TypeError : strlen(): Argument #1 ($string) must be of type string, bool given` when trying to parse a FHIRTiming object with the following input data:

```xml
<timing>
  <repeat>
     <frequency value="1"/>
     <period value="1"/>
     <periodUnit value="d"/>
  </repeat>
</timing>
```

The error is caused by a check if there are decimals. When there are no decimals `strstr` will return false and `strlen(false)` will produce this error. 

https://github.com/dcarbone/php-fhir/blob/master/template/properties/methods/primitive/decimal_type.php#L39

Besides that is frequency of type PositiveInt so a decimal should not be accepted.

**Comment by dcarbone** [comment]
Thanks for reporting this, I'll get it fixed asap.  I'm sure there are more than a few small issues like this throughout this update.

**Comment by dcarbone** [comment]
This should be addressed with #83

**Comment by leonrenkema** [comment]
Thanks for the quick response. I have tested it and it works.

**Comment by leonrenkema** [comment]
Sorry to reopen it but I see some behaviour that is related. 

You default to `$this->_decimals = 1;` and this has a side effect that the generated XML has `1.0` as value when the input is `1`

```php
$doseAndRate->setDoseQuantity((new FHIRQuantity())->setValue("1"));
```

```xml
<doseQuantity>
    <value value="1.0"/>
</doseQuantity>
```

Is this indented behaviour?

**Comment by dcarbone** [comment]
Sure, its a decimal value.  I don't remember specifically where I saw this, but I believe FHIR decimals must always be represented as such.

**Comment by leonrenkema** [comment]
Maybe it does not matter but for me it looks like added precision where it is not there.
And I wonder if it also happens when the type is defined as a positiveInt. 

I can make a testcase for that to check it and how it validates.

**Comment by dcarbone** [comment]
@leonrenkema ah but as mentioned earlier, the `PositiveInt` type is grouped together with other `Integer` types into their own specific primitive type.

You can see the logic that determines which primitive type goes with which class in [this template file](https://github.com/dcarbone/php-fhir/blob/master/template/properties/methods/primitive.php#L60).

**Comment by dcarbone** [comment]
i think the decimal class could do with some improvements in general...

**Comment by leonrenkema** [comment]
@dcarbone I see, the PositiveInt type works with another Primitive type so that is no problem. 

The specification has some notes about precision and the decimal type. https://hl7.org/fhir/datatypes.html#decimal

_The precision of the decimal value has significance:_
- _e.g. 0.010 is regarded as different to 0.01, and the original precision should be preserved_
- _Implementations SHALL handle decimal values in ways that preserve and respect the precision of the value as represented for presentation purposes_

And maybe storing the value as a float is not a good idea? There are differences with precision. My favorite example is `var_dump( 0.1+0.2==0.3)`.

**Comment by dcarbone** [comment]
@leonrenkema I think there are a few things to decide here:

1. If you provide a non-float value to a float register, what is the desired outcome?
   * Personally, I would much prefer a float always be a float.  Sometimes being a float and sometimes being an integer produces ambiguity I am not OK with.
2. The only possible other functional storage medium for this value is a string. This is not an actual solution, however, as when this model is serialized to JSON it _must_ be converted to an actual float value, e.g. not a quoted JSON string, thus we immediately have to consider precision again.

Since the FHIR documentation explicitly states that the precision value is purely for representational purposes, I think one possible acceptable answer is to allow for zero-precision storage.  The tricky part will be to try to find a way to preserve precision when provided a float from the outset, such as would be the case when unserializing from JSON.  I haven't looked into that much with PHP, but I can try.

tldr; I think the only acceptable path forward is:

1. store value as float
2. allow zero precision during serialization
3. implement best-effort precision determination at time of value being set.

**Comment by leonrenkema** [comment]
Maybe an option is to use two integer fields, one for the whole number part and one for the decimal part. Then there is no loss of precision or other floating point "magic".

There is a php decimal type library somewhere. I will take a look into it.

**Comment by dcarbone** [comment]
no, as mentioned ints will not work unless we store three, one for whole number, one for decimal value, and one for precision. as mentioned in their example, we must be able to represent `1.01` and `1.010` as distinct values. 

a decimal library will have the same floating point issues. fp's are simply difficult. and, as mentioned, once serialized from and into json, all of our additional efforts here become less significant.

**Comment by dcarbone** [comment]
Does it serialize it as a string or as a numerical value?

**Comment by leonrenkema** [comment]
It does serialize back to a numerical value, that is interesting. 

![image](https://user-images.githubusercontent.com/497194/202397267-cd371b9f-391a-41b4-aab5-fcb870f23378.png)

Even when I use `"valueDecimal": 42.47000` as input it still keeps the precision. So there is a lot going on with parsing in that library. I am not sure if that can easily be build in PHP with the default json_decode/json_encode.

**Comment by dcarbone** [comment]
Yeah...I'd never before encountered a JSON encoding / decoding library that kept precision past the last significant digit.  I'm not really sure that is something we can "solve" here, especially as I can't imagine it is computationally important to do so...

**Comment by leonrenkema** [comment]
No that is not fixable without another json parser. 

I get it from a scientific perspective, when you send 1.000 you are very sure that is has that precision. It is different from just 1. 

So when sending lab results it can make a difference in theory. But we don't use it and our dosings will not use such a precision.

**Comment by dcarbone** [comment]
@leonrenkema would you be OK closing this issue?  I'm not sure this is worth "fixing" within the realm of PHP until a different json serializing / unserializing mechanism becomes available to us.

**Comment by leonrenkema** [comment]
Yes, this is not easy to fix.
